src/rcc.rs

```diff
@@ -1,26 +1,341 @@
+                rcc.bdcr.modify(|_,w| unsafe {
+                    w.lsedrv()
+                        .bits(0b11)
+                        .lseon()
+                        .set_bit()
+                });
+                while rcc.bdcr.read().lserdy().bit_is_clear() {}
+                rcc.cr.modify(|_,w| w.msipllen().set_bit());
+            }
+            (self.freq(), 0b00)
+        }
+    }
+
+    impl InputClock for MediumSpeedInternalRC {
+        fn freq(&self) -> u32 {
+            self.f
+        }
+    }
+
+    /// High-speed external 4-48 MHz oscillator
+    #[derive(Clone, Copy)]
+    pub struct HighSpeedExternalOSC(pub u32);
+    impl InputClock for HighSpeedExternalOSC {
+        fn freq(&self) -> u32 {
+            self.0
+        }
+    }
+
+    impl HighSpeedExternalOSC {
+        pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+            rcc.cr.modify(|_, w| w
+                          .hseon()
+                          .set_bit()
+                         );
+            while rcc.cr.read().hserdy().bit_is_clear() {}
+            (self.freq(), 0b10)
+        }
+    }
+
+    pub enum RtcClkSource {
+        None,
+        /// Internal 32 kHz RC
+        LSI,
+        /// External 32.768 kHz oscillator
+        LSE,
+        HSEDiv32,
+    }
+    impl RtcClkSource {
+        pub fn freq(&self, hse: Option<HighSpeedExternalOSC>) -> Result<u32, ()> {
+            match *self {
+                RtcClkSource::None => Err(()),
+                RtcClkSource::LSI => Ok(32_000),
+                RtcClkSource::LSE => Ok(32_768),
+                RtcClkSource::HSEDiv32 => {
+                    if let Some(clk) = hse {
+                        Ok(clk.freq() / 32)
+                    } else {
+                        Err(())
+                    }
+                },
+            }
+        }
+
+        pub fn bits(&self) -> u8 {
+            match *self {
+                RtcClkSource::None => 0,
+                RtcClkSource::LSI => 1,
+                RtcClkSource::LSE => 2,
+                RtcClkSource::HSEDiv32 => 3,
+            }
+        }
+
+    }
+
+    #[derive(Clone, Copy)]
+    pub enum SysClkSource {
+        HSI16(HighSpeedInternal16RC),
+        MSI(MediumSpeedInternalRC),
+        HSE(HighSpeedExternalOSC),
+        PLL(PLLClkOutput),
+    }
+    impl InputClock for SysClkSource {
+        fn freq(&self) -> u32 {
+            match *self {
+                SysClkSource::HSI16(_) => 16_000_000,
+                SysClkSource::MSI(s) => s.freq(),
+                SysClkSource::HSE(s) => s.freq(),
+                SysClkSource::PLL(s) => s.freq()
+            }
+        }
+    }
+
+    /// PLLCLK output of PLL module
+    #[derive(Clone, Copy)]
+    pub struct PLLClkOutput {
+        pub src: PLLClkSource,
+        pub m: u8,
+        n: u8,
+        r: u8,
+        f: u32,
+    }
+
+    impl PLLClkOutput {
+        pub fn new(src: PLLClkSource, m: u8, n: u8, r: u8) -> Self {
```

Comment:
Certainly need description and more descriptive names for arguments!

Please add reference to manual's chapter that describes configuration for PLL clock

Reply:
Added a comment - in this case, `m`, `n`, and `r` are really the argument names :D